lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -1030,8 +1129,11 @@ makeChange criteria
     -- Each asset is paired with the complete list of quantities of that asset
     -- present in the selected inputs.
     nonUserSpecifiedAssets :: [(AssetId, NonEmpty TokenQuantity)]
-    nonUserSpecifiedAssets = Map.toList $
-        F.foldr discardUserSpecifiedAssets mempty inputBundles
+    nonUserSpecifiedAssets =
+      let
+        x = Map.toList $ F.foldr discardUserSpecifiedAssets mempty $ ((F.fold inputBundles `TokenBundle.add` TokenBundle.fromTokenMap mintInputs `TokenBundle.difference` TokenBundle.fromTokenMap burnInputs) NE.:| [])
```

Comment:
Hi @sevanspowell 

I think this line is incorrect, unfortunately.

The original function of `nonUserSpecifiedAssets` is to provide a list of all non-user specified assets, where each asset is paired with the complete list of quantities of that asset present in the original inputs.

For example, if a particular non-user-specified asset `a` appears in **_three distinct inputs_**, then asset `a` should be paired with **_three quantities_** in `nonUserSpecifiedAssets`.

However, I _think_ the effect of this change would be to merge these three distinct quantities into a single quantity, which is not what we want (unfortunately).

I'm just writing a property test and set of unit tests to specify the correct behaviour of `nonUserSpecifiedAssets`. I'll leave another comment here when I've done that!

Reply:
Hi @sevanspowell 

If we're going to adjust the definition of `nonUserSpecifiedAssets`, I think it would be a good idea to start from a position where the existing behaviour of this definition is well-defined and tested.

Unfortunately, we haven't really tested this definition in isolation before, and in hindsight, the original code perhaps wasn't as clear as it should have been! (The result is that it's possible to change the definition without causing **_any_** test failures. 😅)

To fix this hole in existing coverage, I've made a PR that tests the original behaviour of `nonUserSpecifiedAssets` by extracting it out into a separate function: https://github.com/input-output-hk/cardano-wallet/pull/2692

I'm not yet sure how we should alter this definition, or even if we should alter it at all. Perhaps we can have a chat in the morning about how we want to proceed!